pkg/datastore/dynamodb/dynamodb.go

```diff
@@ -0,0 +1,302 @@
+			// it will be generated automatically in case of empty string passed.
+			// ref: https://github.com/aws/aws-sdk-go/blob/0dd12669013412980b665d4f6e2947d57b1cd062/aws/credentials/stscreds/web_identity_provider.go#L116-L121
+			stscreds.NewWebIdentityRoleProvider(sts.New(sess), s.roleARN, "", s.tokenFile),
+			&ec2rolecreds.EC2RoleProvider{
+				Client: ec2metadata.New(sess),
+			},
+		},
+	)
+	cfg := aws.NewConfig().WithRegion(region).WithEndpoint(endpoint).WithCredentials(creds)
+	s.client = dynamodb.New(sess, cfg)
+
+	return s, nil
+}
+
+// Find implementation for DynamoDB
+func (s *DynamoDB) Find(ctx context.Context, kind string, opts datastore.ListOptions) (datastore.Iterator, error) {
+	expr, err := buildDynamoDBExpression(opts)
+	if err != nil {
+		s.logger.Error("failed to build query",
+			zap.String("kind", kind),
+			zap.Error(err),
+		)
+	}
+	input := &dynamodb.ScanInput{
+		ExpressionAttributeNames:  expr.Names(),
+		ExpressionAttributeValues: expr.Values(),
+		FilterExpression:          expr.Filter(),
+		TableName:                 aws.String(kind),
+	}
+	// TODO: Support pagination with ListOptions.Page and ListOptions.PageSize for DynamoDB.
+	// TODO: Support ordering with ListOptions.Orders for DynamoDB.
+	var items []map[string]*dynamodb.AttributeValue
+	err = s.client.ScanPagesWithContext(ctx, input,
+		func(page *dynamodb.ScanOutput, lastPage bool) bool {
+			items = append(items, page.Items...)
+			// The only way to know when you have reached
+			// the end of the result set is when LastEvaluatedKey is empty.
+			return len(page.LastEvaluatedKey) != 0
+		},
+	)
+	if err != nil {
+		s.logger.Error("failed to get cursor",
+			zap.String("kind", kind),
+			zap.Error(err),
+		)
+		return nil, fmt.Errorf("failed to get cursor: %v", err)
+	}
+	return &Iterator{
+		datapool: items,
+	}, nil
+}
+
+// Get implementation for DynamoDB
+func (s *DynamoDB) Get(ctx context.Context, kind, id string, v interface{}) error {
+	input := &dynamodb.GetItemInput{
+		TableName: aws.String(kind),
+		Key: map[string]*dynamodb.AttributeValue{
+			"Id": {
+				S: aws.String(id),
+			},
+		},
+	}
+	result, err := s.client.GetItemWithContext(ctx, input)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case dynamodb.ErrCodeResourceNotFoundException:
+				return datastore.ErrNotFound
+			default:
+				s.logger.Error("failed to retrieve entity: aws error",
+					zap.String("id", id),
+					zap.String("kind", kind),
+					zap.Error(err),
+				)
+				return err
+			}
+		}
+		s.logger.Error("failed to retrieve entity",
+			zap.String("id", id),
+			zap.String("kind", kind),
+			zap.Error(err),
+		)
+		return err
+	}
+
+	return dynamodbattribute.UnmarshalMap(result.Item, v)
+}
+
+// Create implementation for DynamoDB
+func (s *DynamoDB) Create(ctx context.Context, kind, id string, entity interface{}) error {
+	err := s.Get(ctx, kind, id, entity)
+	if err == nil {
+		return datastore.ErrAlreadyExists
```

Comment:
Don't we need a transaction to avoid adding the same ones?

Reply:
Sorry I could not get your point 👀 currently, in `Create` function we use `Get` first to check the existence and only create/put in case of no the same key existed in the database. So where should we use transaction for this atomic action (create/put a row to one table)?

Reply:
`Create` means adding the entity only if it is NOT existing.
`Update` means updating the entity only if it is existing.
`Put` means adding or overriding the entity (this does not care about the existence of entity)

This `Create` or `Update` function can be called simultaneously from multiple callers.
So we have to ensure that the check of entity existence and updating must be atomic.

If not they will enter an unexpected behavior in some cases. For example:
- it checks the existence by using `Get` and ensure that no entity was found
- before adding the entity into DB, the same entity was added by another caller

I am not sure about DynamoDB but from its docs, it seems we can accomplish the goal by using transaction or using [conditional writes](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithItems.html#WorkingWithItems.ReadingData).

You can also check out the implementation for Firestore [here](https://github.com/pipe-cd/pipe/blob/dynamodb-support/pkg/datastore/firestore/firestore.go).

Reply:
> This Create or Update function can be called simultaneously from multiple callers.

Ah, I see 👍 That the point. In that case, using GetItemInput.ConsistentRead parameter for Get operation may be enough
```
	// Determines the read consistency model: If set to true, then the operation
	// uses strongly consistent reads; otherwise, the operation uses eventually
	// consistent reads.
	ConsistentRead *bool `type:"boolean"`
```
What do you think? 🤔

Reply:
Thanks for your comment 👍 
Will resolve this consistency manipulate operations issue with [ConditionExpresions](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ConditionExpressions.html) (attribute_exists | attribute_not_exists function). 🙏